### ISSUE TITLE
Google compute engine instance and disk referencing error

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-google#112]
_This issue was originally opened by @nodesocket as hashicorp/terraform#15181. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Think I discovered a bug using `google_compute_disk` and `google_compute_instance` referencing resources.

Create a disk using a resource block like:

```
resource "google_compute_disk" "blog" {
    name  = "blog"
    type  = "pd-ssd"
    zone  = "us-west1-a"
    snapshot = "some-snapshot-here"
    size = 10
}
```

Then create an instance referencing the above disk:

```
resource "google_compute_instance" "blog" {
    name = "blog"
    machine_type = "f1-micro"
    zone = "${google_compute_disk.blog.zone}"

    tags = []

    disk {
        disk = "${google_compute_disk.blog.name}"
        auto_delete = true
    }

    network_interface {
        network = "default"
        access_config {
            # ephemeral external ip address
        }
    }

    metadata {
    }

    service_account {
        scopes = ["https://www.googleapis.com/auth/servicecontrol",
                  "https://www.googleapis.com/auth/service.management.readonly",
                  "https://www.googleapis.com/auth/logging.write",
                  "https://www.googleapis.com/auth/monitoring.write",
                  "https://www.googleapis.com/auth/trace.append",
                  "https://www.googleapis.com/auth/devstorage.read_only",
                  "https://www.googleapis.com/auth/cloud.useraccounts.readonly"]
    }

    scheduling {
        preemptible = false
        on_host_maintenance = "MIGRATE"
        automatic_restart = true
    }
}
```

After the `apply` finishes, run `apply` again. You will see the following error:

```
Error applying plan:

1 error(s) occurred:

* google_compute_disk.blog (destroy): 1 error(s) occurred:

* google_compute_disk.blog: Error detaching disk persistent-disk-0 from instance project-id-here/https://www.googleapis.com/compute/v1/projects/project-id-here/zones/us-west1-a/blog: googleapi: Error 400: Invalid value 'https://www.googleapis.com/compute/v1/projects/project-id-here/zones/us-west1-a'. Values must match the following regular expression: '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?', invalidParameter

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by nodesocket** [comment]
Bump of this please. The specific error seems to be:

``bash
Error detaching disk persistent-disk-0 from instance project-id-here/https://www.googleapis.com/compute/v1/projects/project-id-here/zones/us-west1-a/blog: googleapi: Error 400: Invalid value 'https://www.googleapis.com/compute/v1/projects/project-id-here/zones/us-west1-a'. Values must match the following regular expression: '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?', invalidParameter
```

**Comment by danawillow** [comment]
Hey @nodesocket, sorry you're having issues. I'm actually doing a big rework of the disk attaching process so I'm not going to be able to get to this specific issue right away, but it should be resolved reasonably soon as a result of the other work!

**Comment by nodesocket** [comment]
Great. Is this specific to creating a disk from a `snapshot` or does this bug also apply to creating disks via an `image` and associating with a VM instance?

**Comment by danawillow** [comment]
My work is focusing on the latter.

**Comment by nodesocket** [comment]
Ok, well I need both cases to work 😸 .

Create disk from snapshot

```
resource "google_compute_disk" "blog" {
    name  = "blog"
    type  = "pd-ssd"
    zone  = "us-west1-a"
    snapshot = "some-snapshot-here"
    size = 10
}
```

Create disk from image

```
resource "google_compute_disk" "blog" {
    name  = "blog"
    type  = "pd-ssd"
    zone  = "us-west1-a"
    image = "some-image-here"
    size = 10
}
```

```
disk {
    disk = "${google_compute_disk.blog.name}"
    auto_delete = true
}
```

**Comment by nodesocket** [comment]
Any update on this? Seems to be still a pretty big bug creating a GCP disk from a snapshot. I believe creating a disk from an image _(inline in the instance)_ works though.

**Comment by EssentialMix** [comment]
Can confirm that this is still a issue and related to disk created using snapshot, terraform all time want to recreate disks created using snapshot:
terraform plan
-/+ google_compute_disk.web-disk01
    disk_encryption_key_sha256: "" => "<computed>"
    name:                       "test-01-prod-us-central1-b-disk" => "test-01-prod-us-central1-b-disk"
    self_link:                  "https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1-b/disks/test-01-prod-us-central1-b-disk" => "<computed>"
    size:                       "1000" => "1000"
    snapshot:                   "https://www.googleapis.com/compute/v1/projects/my-project/global/snapshots/test-disk-snapshot-test" => "test-disk-snapshot-test" (forces new resource)
    type:                       "pd-ssd" => "pd-ssd"
    users.#:                    "1" => "<computed>"
    zone:                       "us-central1-b" => "us-central1-b"


Plan: 1 to add, 0 to change, 1 to destroy.

**Comment by EssentialMix** [comment]
BTW if snapshot set to:
`https://www.googleapis.com/compute/v1/projects/my-project/global/snapshots/test-disk-snapshot-test`

instead of name: test-disk-snapshot-test which, which was used originally during instance creation, it will work

**Comment by nodesocket** [comment]
This is fixed in `0.10.0`. Note, the syntax slightly changed from `0.9.6`.

````
resource "google_compute_disk" "blog" {
    name = "blog"
    type = "pd-ssd"
    zone = "${var.blog_zone}"
    snapshot = "blog-2017-07-12-1"
    size = 10
}

resource "google_compute_instance" "blog" {
    ....
    boot_disk {
        source = "${google_compute_disk.blog.name}"
        auto_delete = true
    }
    ....
````

**Comment by rileykarson** [comment]
The disk should be successfully detached with #320; I just wanted to note that this doesn't mean that attaching and detaching will work, and that deleting a disk means it will be removed completely from the instance. The delete operation will just succeed now, instead of failing unintentionally.
